linalg/pardiso.cpp

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2010-2023, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "pardiso.hpp"
+
+#ifdef MFEM_USE_MKL_PARDISO
+
+#include "sparsemat.hpp"
+
+namespace mfem
+{
+
+PardisoSolver::PardisoSolver()
+{
+   // Indicate that default parameters are changed
+   iparm[0] = 1;
+   // Use METIS for fill-in reordering
+   iparm[1] = 2;
+   // Write the solution into the x vector data
+   iparm[5] = 0;
+   // Maximum number of iterative refinement steps
+   iparm[7] = 2;
+   // Perturb the pivot elements with 1E-13
+   iparm[9] = 13;
+   // Use nonsymmetric permutation
+   iparm[10] = 1;
+   // Perform a check on the input data
+   iparm[26] = 1;
+   // 0-based indexing in CSR data structure
+   iparm[34] = 1;
+   // Maximum number of numerical factorizations
+   maxfct = 1;
+   // Which factorization to use. This parameter is ignored and always assumed
+   // to be equal to 1. See MKL documentation.
+   mnum = 1;
+   // Print statistical information in file
+   msglvl = 0;
+   // Initialize error flag
+   error = 0;
+   // Real nonsymmetric matrix
+   mtype = MatType::REAL_NONSYMMETRIC;
+   // Number of right hand sides
+   nrhs = 1;
+}
+
+void PardisoSolver::SetOperator(const Operator &op)
+{
+   auto mat = const_cast<SparseMatrix *>(dynamic_cast<const SparseMatrix *>(&op));
+
+   MFEM_ASSERT(mat, "Must pass SparseMatrix as Operator");
+
+   m = mat->Size();
+
+   nnz = mat->NumNonZeroElems();
+
+   if (mat->IsSymmetric() == 0.0)
```

Comment:
This check might be unnecessarily expensive since it will parse all the entries of the matrix to check for symmetry. 
Maybe allowing the user to set the symmetry would suffice.

Reply:
I agree to remove this

Reply:
Same: leave it to the use to specify if symmetric, good catch @psocratis